modules/consensus/tendermint/verifier/src/verifier.rs

```diff
@@ -0,0 +1,355 @@
+		.map_err(|e| VerificationError::Invalid(e.to_string()))?;
+	let timestamp = Timestamp { seconds: trusted_state.timestamp as i64, nanos: 0 };
+	let time = Time::try_from(timestamp).map_err(|e| VerificationError::Invalid(e.to_string()))?;
+	let trusted_next_validators = ValidatorSet::new(trusted_state.next_validators.clone(), None);
+	let trusted_next_validators_hash = Hash::Sha256(trusted_state.next_validators_hash);
+
+	let tendermint_trusted_state = TrustedBlockState {
+		chain_id: &chain_id,
+		header_time: time,
+		height,
+		next_validators: &trusted_next_validators,
+		next_validators_hash: trusted_next_validators_hash,
+	};
+
+	let untrusted_block_state = UntrustedBlockState {
+		signed_header: &consensus_proof.signed_header,
+		validators: &validators,
+		next_validators: next_validators.as_ref(),
+	};
+
+	let verifier_options = convert_verification_options(
+		&trusted_state.verification_options,
+		trusted_state.trusting_period_duration(),
+	)?;
+	let now = convert_timestamp(current_time)?;
+
+	let verifier = SpIoVerifier::default();
+
+	validate_ancestry_chain(&consensus_proof, &trusted_state)?;
+
+	let result = verifier.verify_misbehaviour_header(
+		untrusted_block_state,
+		tendermint_trusted_state,
+		&verifier_options,
+		now,
+	);
+
+	match result {
+		Verdict::Success => {
+			let updated_state = create_updated_trusted_state(&trusted_state, &consensus_proof)?;
+			Ok(updated_state)
+		},
+		Verdict::NotEnoughTrust(tally) =>
+			Err(VerificationError::NotEnoughTrust(format!("Voting power tally: {}", tally))),
+		Verdict::Invalid(detail) => Err(VerificationError::Invalid(format!("{:?}", detail))),
+	}
+}
+
+/// Validates which validator set the signed header references and check next_validators_hash
+/// rotation. Returns the correct ValidatorSet for the header and validates next_validators if
+/// rotation is signaled.
+fn extract_validators<'a>(
+	trusted_state: &'a TrustedState,
+	consensus_proof: &'a ConsensusProof,
+) -> Result<ValidatorSet, VerificationError> {
+	let signed_header_validators_hash = &consensus_proof.signed_header.header.validators_hash;
+	let current_validators_hash =
+		ValidatorSet::new(trusted_state.validators.clone(), None).hash_with::<SpIoSha256>();
+	let next_validators_hash = Hash::Sha256(trusted_state.next_validators_hash);
+
+	let validators = if signed_header_validators_hash == &current_validators_hash {
+		ValidatorSet::new(trusted_state.validators.clone(), None)
+	} else if signed_header_validators_hash == &next_validators_hash {
+		ValidatorSet::new(trusted_state.next_validators.clone(), None)
+	} else {
+		return Err(VerificationError::Invalid(format!(
+			"Unknown validator set hash: {:?}",
+			signed_header_validators_hash
+		)));
+	};
+
+	// Validate next_validators_hash rotation if signaled
+	let signed_header_next_validators_hash =
+		&consensus_proof.signed_header.header.next_validators_hash;
+	if signed_header_next_validators_hash != &next_validators_hash {
```

Comment:
This check should only run when `signed_header_next_validators_hash` is not empty, From the cometbft go code,  next_validator_hash is allowed to be empty.
https://github.com/cometbft/cometbft/blob/bd1bc3bb6392a973d7b60fd0b19da6e71ca09b47/types/block.go#L426

https://github.com/cometbft/cometbft/blob/bd1bc3bb6392a973d7b60fd0b19da6e71ca09b47/types/validation.go#L247